commit 2314ad56620e
Author: Randell Jesup <rjesup@jesup.org>
Date:   Fri May 25 21:16:28 2018 -0400

    Bug 1463494 - Delete the sensor observerlist array in a deferred manner. r=froydnj, r=jchen, a=abillings
    
    Clean up sensorlist if Dispatch fails.
    
    --HG--
    extra : source : 494d77b62c333c2b71097d4ecf41c1a60dc495a2
---
 hal/Hal.cpp | 19 +++++++++++++++++--
 1 file changed, 17 insertions(+), 2 deletions(-)

diff --git hal/Hal.cpp hal/Hal.cpp
index ddd7121a91dd1..6b389407f737a 100644
--- hal/Hal.cpp
+++ hal/Hal.cpp
@@ -439,14 +439,29 @@ UnregisterSensorObserver(SensorType aSensor, ISensorObserver *aObserver) {
   }
   DisableSensorNotifications(aSensor);
 
-  // Destroy sSensorObservers only if all observer lists are empty.
   for (int i = 0; i < NUM_SENSOR_TYPE; i++) {
     if (gSensorObservers[i].Length() > 0) {
       return;
     }
   }
-  delete [] gSensorObservers;
+
+  // We want to destroy gSensorObservers if all observer lists are
+  // empty, but we have to defer the deallocation via a runnable to
+  // mainthread (since we may be inside NotifySensorChange()/Broadcast()
+  // when it calls UnregisterSensorObserver()).
+  SensorObserverList* sensorlists = gSensorObservers;
   gSensorObservers = nullptr;
+
+  // Unlike DispatchToMainThread, DispatchToCurrentThread doesn't leak a runnable if
+  // it fails (and we assert we're on MainThread).
+  if (NS_FAILED(NS_DispatchToCurrentThread(NS_NewRunnableFunction("UnregisterSensorObserver",
+                                                                  [sensorlists]() -> void {
+      delete [] sensorlists;
+      }))))
+  {
+    // Still need to delete sensorlists if the dispatch fails
+    delete [] sensorlists;
+  }
 }
 
 void
